presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java

```diff
@@ -281,6 +285,20 @@ public boolean needsInput()
 
     @Override
     public void addInput(Page page)
+    {
+        addPage(page);
+        // Calculating the retained size of all page builders in partitionFunction.getRetainedSizeInBytes()
+        // can be expensive especially for complex types. Therefore, we update the memory usage no
+        // more than once per second.
+        long now = System.nanoTime();
+        if (now - lastMemoryUsageUpdateNanos > ONE_SECOND) {
```

Comment:
Instead of updating the memory periodically, I'm thinking updating the memory when it needs to change (https://github.com/highker/presto/commit/a618a9b8e69e30664b3e34fa75d99a7476a2eb2c). Or something along the line. There should be a cheap way to avoid necessary update of memory.

Reply:
Does that solution improve this simple approach in some way? Also the condition `if (flushSizeInBytes > partitionBuilderRetainedSizes[partition])` will cause the used memory to be updated only if there is an increase, so we won't capture the downward trends if there is any. Finally, it has a downside of being slightly more complicated. 

One thing to try is, in `flush()` when the condition `if (!partitionPageBuilder.isEmpty() && (force || partitionPageBuilder.isFull()))` holds we can update the memory usage (remove the `if` on L442). But, the overhead of calculating the memory usage will depend on how frequently page builders get full, which will make it somewhat unpredictable.

The reason I went with this approach is it's simple and its overhead is predictable.